cmd/ipfs/pinmfs.go

```diff
@@ -0,0 +1,252 @@
+			return
+		case <-tmo.C:
+		}
+
+		// reread the config, which may have changed in the meantime
+		cfg, err := cctx.GetConfigNoCache()
+		if err != nil {
+			mfslog.Errorf("pinning reading config (%v)", err)
+			select {
+			case errCh <- err:
+			case <-cctx.Context().Done():
+				return
+			}
+			continue
+		}
+		mfslog.Debugf("pinning loop is awake, %d remote services", len(cfg.Pinning.RemoteServices))
+
+		// get the most recent MFS root cid
+		rootNode, err := node.RootNode()
+		if err != nil {
+			mfslog.Errorf("pinning reading MFS root (%v)", err)
+			select {
+			case errCh <- err:
+			case <-cctx.Context().Done():
+				return
+			}
+			continue
+		}
+		rootCid := rootNode.Cid()
+
+		// pin to all remote services in parallel
+		pinAllMFS(cctx.Context(), node, cfg, rootCid, lastPins, errCh)
+	}
+}
+
+// pinAllMFS pins on all remote services in parallel to overcome DoS attacks.
+func pinAllMFS(ctx context.Context, node pinMFSNode, cfg *config.Config, rootCid cid.Cid, lastPins map[string]lastPin, errCh chan<- error) {
+	ch := make(chan lastPin, len(cfg.Pinning.RemoteServices))
+	for svcName_, svcConfig_ := range cfg.Pinning.RemoteServices {
+		// skip services where MFS is not enabled
+		svcName, svcConfig := svcName_, svcConfig_
+		mfslog.Debugf("pinning considering service %s for mfs pinning", svcName)
+		if !svcConfig.Policies.MFS.Enable {
+			mfslog.Debugf("pinning service %s is not enabled", svcName)
+			ch <- lastPin{}
+			continue
+		}
+		// read mfs pin interval for this service
+		var repinInterval time.Duration
+		if svcConfig.Policies.MFS.RepinInterval == "" {
+			repinInterval = defaultRepinInterval
+		} else {
+			var err error
+			repinInterval, err = time.ParseDuration(svcConfig.Policies.MFS.RepinInterval)
+			if err != nil {
+				mfslog.Errorf("pinning parsing service %s repin interval %q", svcName, svcConfig.Policies.MFS.RepinInterval)
+				select {
+				case errCh <- fmt.Errorf("remote pinning service %s has invalid MFS.RepinInterval (%v)", svcName, err):
+				case <-ctx.Done():
+				}
+				ch <- lastPin{}
+				continue
+			}
+		}
+
+		// do nothing, if MFS has not changed since last pin on the exact same service or waiting for MFS.RepinInterval
+		if last, ok := lastPins[svcName]; ok {
+			if last.ServiceConfig == svcConfig && (last.CID == rootCid || time.Since(last.Time) < repinInterval) {
+				if last.CID == rootCid {
+					mfslog.Debugf("pinning MFS root to %s: pin for %s exists since %s, skipping", svcName, rootCid, last.Time.String())
+				} else {
+					mfslog.Debugf("pinning MFS root to %s: skipped due to MFS.RepinInterval=%s (remaining: %s)", svcName, repinInterval.String(), (repinInterval - time.Since(last.Time)).String())
+				}
+				ch <- lastPin{}
+				continue
+			}
+		}
+
+		mfslog.Debugf("pinning MFS root %s to %s", rootCid, svcName)
+		go func() {
+			if r, err := pinMFS(ctx, node, rootCid, svcName, svcConfig); err != nil {
+				mfslog.Debugf("pinning MFS root %s to %s error (%v)", rootCid, svcName, err)
```

Comment:
Any reason why we not make it `ERROR` instead? Will make it extra clear when service has problems, and avoid people filling issues against go-ipfs when pinning fails due to  remote service  error:

```suggestion
				mfslog.Errorf("pinning MFS root %s to %s error (%v)", rootCid, svcName, err)
```